File: apps/programacao/src/lib/dominio/conflitos.ts
import { TIPO_RECURSO } from './constantes'

/**
 * O que está errado no quadro — conferido em função pura, sem banco.
 *
 * Existe porque é justamente isto que o Excel não faz. Hoje o mesmo carro entra em duas
 * frentes e ninguém percebe até de manhã, com duas turmas esperando o mesmo Doblô; e um
 * veículo que entrou na oficina continua sendo escalado porque quem monta a programação à
 * noite não viu a manutenção aberta de manhã.
 *
 * Tudo aqui é AVISO, não impedimento. Escalar o mesmo carro em duas frentes pode ser
 * proposital (leva a turma e volta), e travar o quadro faria o chefe voltar para o Excel,
 * onde nada trava. Mostrar e deixar decidir é o que mantém a ferramenta em uso.
 */

export type Gravidade = 'alta' | 'media'

export type Conflito = {
  gravidade: Gravidade
  titulo: string
  detalhe: string
  frenteIds: string[]
}

type EscalaMin = { id: string; frenteId: string; nome: string; funcaoSigla: string | null }
type RecursoMin = {
  id: string; frenteId: string; tipo: string
  placa: string | null; descricao: string; motoristaNome: string | null
}
type FrenteMin = { id: string; nome: string }
type VeiculoMin = { placa: string; emManutencao: boolean; motivoManutencao: string | null }

function normalizar(t: string): string {
  return t.trim().toLowerCase().normalize('NFD').replace(/[̀-ͯ]/g, '')
}

export function conferirQuadro(
  escalas: EscalaMin[],
  recursos: RecursoMin[],
  frentes: FrenteMin[],
  veiculosDaFrota: VeiculoMin[],
): Conflito[] {
  const nomeDaFrente = new Map(frentes.map((f) => [f.id, f.nome]))
  const conflitos: Conflito[] = []

  // ── Mesma pessoa em duas frentes ───────────────────────────────────────────
  const frentesPorPessoa = new Map<string, Set<string>>()
  for (const e of escalas) {
    const chave = normalizar(e.nome)
    if (!frentesPorPessoa.has(chave)) frentesPorPessoa.set(chave, new Set())
    frentesPorPessoa.get(chave)!.add(e.frenteId)
  }
  for (const [chave, ids] of frentesPorPessoa) {
    if (ids.size < 2) continue
    const nome = escalas.find((e) => normalizar(e.nome) === chave)!.nome
    conflitos.push({
      gravidade: 'alta',
      titulo: `${nome} está em ${ids.size} frentes`,
      detalhe: [...ids].map((i) => nomeDaFrente.get(i) ?? '?').join(' e '),
      frenteIds: [...ids],
    })
  }

  // ── Mesmo veículo em duas frentes ──────────────────────────────────────────
  const frentesPorPlaca = new Map<string, Set<string>>()
  for (const r of recursos) {
    if (r.tipo !== TIPO_RECURSO.VEICULO || !r.placa) continue
    const chave = r.placa.replace(/\W/g, '').toUpperCase()
    if (!frentesPorPlaca.has(chave)) frentesPorPlaca.set(chave, new Set())
    frentesPorPlaca.get(chave)!.add(r.frenteId)
  }
  for (const [placa, ids] of frentesPorPlaca) {
    if (ids.size < 2) continue
    conflitos.push({
      gravidade: 'alta',
      titulo: `Veículo ${placa} em ${ids.size} frentes`,
      detalhe: [...ids].map((i) => nomeDaFrente.get(i) ?? '?').join(' e '),
      frenteIds: [...ids],
    })
  }

  // ── Veículo com manutenção aberta ──────────────────────────────────────────
  const naOficina = new Map(
    veiculosDaFrota.filter((v) => v.emManutencao).map((v) => [v.placa.replace(/\W/g, '').toUpperCase(), v]),
  )
  for (const r of recursos) {
    if (r.tipo !== TIPO_RECURSO.VEICULO || !r.placa) continue
    const v = naOficina.get(r.placa.replace(/\W/g, '').toUpperCase())
    if (!v) continue
    conflitos.push({
      gravidade: 'alta',
      titulo: `${r.descricao} está em manutenção`,
      detalhe: `${nomeDaFrente.get(r.frenteId) ?? '?'} · ${v.motivoManutencao ?? 'manutenção aberta na Frota'}`,
      frenteIds: [r.frenteId],
    })
  }

  // ── Motorista escalado em outra frente ─────────────────────────────────────
  for (const r of recursos) {
    if (!r.motoristaNome?.trim()) continue
    const alvo = normalizar(r.motoristaNome)
    const ondeEsta = escalas.filter((e) => normalizar(e.nome) === alvo).map((e) => e.frenteId)
    if (ondeEsta.length === 0) {
      conflitos.push({
        gravidade: 'media',
        titulo: `${r.motoristaNome} dirige mas não está escalado`,
        detalhe: `${r.descricao} · ${nomeDaFrente.get(r.frenteId) ?? '?'}`,
        frenteIds: [r.frenteId],
      })
      continue
    }
    if (!ondeEsta.includes(r.frenteId)) {
      conflitos.push({
        gravidade: 'media',
        titulo: `${r.motoristaNome} dirige em uma frente e trabalha em outra`,
        detalhe:
          `${r.descricao} está em ${nomeDaFrente.get(r.frenteId) ?? '?'}, ` +
          `e ele está escalado em ${ondeEsta.map((i) => nomeDaFrente.get(i) ?? '?').join(', ')}`,
        frenteIds: [r.frenteId, ...ondeEsta],
      })
    }
  }

  // ── Veículo sem motorista ──────────────────────────────────────────────────
  for (const r of recursos) {
    if (r.tipo !== TIPO_RECURSO.VEICULO || r.motoristaNome?.trim()) continue
    conflitos.push({
      gravidade: 'media',
      titulo: `${r.descricao} sem motorista`,
      detalhe: nomeDaFrente.get(r.frenteId) ?? '?',
      frenteIds: [r.frenteId],
    })
  }

  const peso = { alta: 0, media: 1 }
  return conflitos.sort((a, b) => peso[a.gravidade] - peso[b.gravidade])
}

File: migracao-go/internal/domain/programacao/conflitos.go
package programacao

import (
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// Conflito é o que está errado no quadro — função pura, sem banco. Existe porque é
// justamente isto que o Excel/WhatsApp não fazem: o mesmo carro entra em duas frentes e
// ninguém percebe até de manhã, com duas turmas esperando o mesmo veículo.
//
// Tudo aqui é AVISO, não impedimento. Escalar o mesmo carro em duas frentes pode ser
// proposital (leva a turma e volta) — travar o quadro faria o chefe voltar pro Excel, onde
// nada trava. Mostrar e deixar decidir é o que mantém a ferramenta em uso.
type Gravidade string

const (
	GravidadeAlta  Gravidade = "alta"
	GravidadeMedia Gravidade = "media"
)

type Conflito struct {
	Gravidade Gravidade
	Titulo    string
	Detalhe   string
	FrenteIDs []string
}

var reNaoAlfanumerico = regexp.MustCompile(`[^a-zA-Z0-9]`)

func normalizar(t string) string {
	return strings.ToLower(strings.TrimSpace(t))
}

func somentePlaca(p string) string {
	return strings.ToUpper(reNaoAlfanumerico.ReplaceAllString(p, ""))
}

// ConferirQuadro recebe as escalas e recursos já carregados de um dia e devolve os
// conflitos encontrados, mais graves primeiro. veiculosEmManutencao é opcional (a Frota
// segue fora da migração — ver README "Frota fora do escopo"); passe nil quando não tiver
// como consultar.
func ConferirQuadro(escalas []Escala, recursos []Recurso, frentes []Frente, veiculosEmManutencao map[string]string) []Conflito {
	nomeDaFrente := map[string]string{}
	for _, f := range frentes {
		nomeDaFrente[f.ID] = f.Nome
	}
	rotulo := func(id string) string {
		if n, ok := nomeDaFrente[id]; ok {
			return n
		}
		return "?"
	}
	juntar := func(ids []string) string {
		nomes := make([]string, len(ids))
		for i, id := range ids {
			nomes[i] = rotulo(id)
		}
		return strings.Join(nomes, " e ")
	}

	var conflitos []Conflito

	// ── Mesma pessoa em duas frentes ───────────────────────────────────────────
	frentesPorPessoa := map[string]map[string]bool{}
	nomeOriginal := map[string]string{}
	ordemPessoa := []string{}
	for _, e := range escalas {
		chave := normalizar(e.Nome)
		if frentesPorPessoa[chave] == nil {
			frentesPorPessoa[chave] = map[string]bool{}
			nomeOriginal[chave] = e.Nome
			ordemPessoa = append(ordemPessoa, chave)
		}
		frentesPorPessoa[chave][e.FrenteID] = true
	}
	for _, chave := range ordemPessoa {
		ids := chavesOrdenadas(frentesPorPessoa[chave])
		if len(ids) < 2 {
			continue
		}
		conflitos = append(conflitos, Conflito{
			Gravidade: GravidadeAlta,
			Titulo:    nomeOriginal[chave] + " está em " + strconv.Itoa(len(ids)) + " frentes",
			Detalhe:   juntar(ids),
			FrenteIDs: ids,
		})
	}

	// ── Mesmo veículo em duas frentes ──────────────────────────────────────────
	frentesPorPlaca := map[string]map[string]bool{}
	placaOriginal := map[string]string{}
	ordemPlaca := []string{}
	for _, r := range recursos {
		if r.Tipo != RecursoVeiculo || r.Placa == nil || *r.Placa == "" {
			continue
		}
		chave := somentePlaca(*r.Placa)
		if frentesPorPlaca[chave] == nil {
			frentesPorPlaca[chave] = map[string]bool{}
			placaOriginal[chave] = chave
			ordemPlaca = append(ordemPlaca, chave)
		}
		frentesPorPlaca[chave][r.FrenteID] = true
	}
	for _, chave := range ordemPlaca {
		ids := chavesOrdenadas(frentesPorPlaca[chave])
		if len(ids) < 2 {
			continue
		}
		conflitos = append(conflitos, Conflito{
			Gravidade: GravidadeAlta,
			Titulo:    "Veículo " + placaOriginal[chave] + " em " + strconv.Itoa(len(ids)) + " frentes",
			Detalhe:   juntar(ids),
			FrenteIDs: ids,
		})
	}

	// ── Veículo com manutenção aberta na Frota ─────────────────────────────────
	for _, r := range recursos {
		if r.Tipo != RecursoVeiculo || r.Placa == nil || *r.Placa == "" {
			continue
		}
		motivo, emManutencao := veiculosEmManutencao[somentePlaca(*r.Placa)]
		if !emManutencao {
			continue
		}
		if motivo == "" {
			motivo = "manutenção aberta na Frota"
		}
		conflitos = append(conflitos, Conflito{
			Gravidade: GravidadeAlta,
			Titulo:    r.Descricao + " está em manutenção",
			Detalhe:   rotulo(r.FrenteID) + " · " + motivo,
			FrenteIDs: []string{r.FrenteID},
		})
	}

	// ── Motorista escalado em outra frente (ou não escalado) ───────────────────
	for _, r := range recursos {
		if r.MotoristaNome == nil || strings.TrimSpace(*r.MotoristaNome) == "" {
			continue
		}
		alvo := normalizar(*r.MotoristaNome)
		var ondeEsta []string
		for _, e := range escalas {
			if normalizar(e.Nome) == alvo {
				ondeEsta = append(ondeEsta, e.FrenteID)
			}
		}
		if len(ondeEsta) == 0 {
			conflitos = append(conflitos, Conflito{
				Gravidade: GravidadeMedia,
				Titulo:    *r.MotoristaNome + " dirige mas não está escalado",
				Detalhe:   r.Descricao + " · " + rotulo(r.FrenteID),
				FrenteIDs: []string{r.FrenteID},
			})
			continue
		}
		if !contem(ondeEsta, r.FrenteID) {
			conflitos = append(conflitos, Conflito{
				Gravidade: GravidadeMedia,
				Titulo:    *r.MotoristaNome + " dirige em uma frente e trabalha em outra",
				Detalhe:   r.Descricao + " está em " + rotulo(r.FrenteID) + ", e ele está escalado em " + juntar(ondeEsta),
				FrenteIDs: append([]string{r.FrenteID}, ondeEsta...),
			})
		}
	}

	// ── Veículo sem motorista ──────────────────────────────────────────────────
	for _, r := range recursos {
		if r.Tipo != RecursoVeiculo {
			continue
		}
		if r.MotoristaNome != nil && strings.TrimSpace(*r.MotoristaNome) != "" {
			continue
		}
		conflitos = append(conflitos, Conflito{
			Gravidade: GravidadeMedia,
			Titulo:    r.Descricao + " sem motorista",
			Detalhe:   rotulo(r.FrenteID),
			FrenteIDs: []string{r.FrenteID},
		})
	}

	return ordenarPorGravidade(conflitos)
}

func chavesOrdenadas(m map[string]bool) []string {
	xs := make([]string, 0, len(m))
	for k := range m {
		xs = append(xs, k)
	}
	return xs
}
func contem(xs []string, alvo string) bool {
	for _, x := range xs {
		if x == alvo {
			return true
		}
	}
	return false
}
func ordenarPorGravidade(cs []Conflito) []Conflito {
	peso := map[Gravidade]int{GravidadeAlta: 0, GravidadeMedia: 1}
	out := make([]Conflito, len(cs))
	copy(out, cs)
	sort.SliceStable(out, func(i, j int) bool { return peso[out[i].Gravidade] < peso[out[j].Gravidade] })
	return out
}
